proposed/verifiable-ref-returns.md

```diff
@@ -0,0 +1,75 @@
+
+NOTE: C# rules operate with high level language entities such as scopes, expressions, local variables, etc., that do not exist at the level of IL or do not map one-to-one.
+The following is the attempt to introduce/adjust verification rules to allow validation of the safety of the ref returns at IL level.
+
+# Verification of ref returns. #
+
+## Returnable managed pointer. ##
+
+CLI needs to add a notion of another transient type - returnable managed pointer. 
+*(need to adjust �III.1.8.1.1 Verification algorithm� and in �I.8.7 Assignment compatibility� accordingly)*
+
+NOTE: there are similarities with already existing concept of controlled mutability managed references.
+
+Returnable managed pointer is a transient type used for verification purposes. When certain conditions are met, managed pointers can be classified as returnable. Unlike ordinary managed pointer types, returnable managed pointers can be used as operands of `RET` instruction in verifiable code. 
+
+Returnable managed pointer may be a result of one of the following:
+-	`LDSFLDA`, `LDELEMA`, `UNBOX` always returnable   (see: C# rule #1)
+-	`LDFLDA` when the receiver is a reference type     (see: C# rule #1).
+-	`LDFLDA` when the receiver is a returnable ref 	(see: C# rule #4)
+-	`LDARG` of a byref parameter 			(see: C# rule #2, #3), 
+except for arg0 in a struct method 		(see: C# rule #5).
+-	`CALL`, `CALLVIRT` when all reference arguments are returnable. (see: C# rule 6)
+note: arg0 in this-calls are not considered here since it cannot be returned by reference.
+-	`LDLOC` of a byref local when the slot marked �returnable� (IL specific rule)
+
+All other ways of obtaining a managed reference (Ex: `LDARGA`, `LDLOCA`, `LOCALLOC`, `REFANYVAL`�) do not result in returnable references.
+
+## Merging stack states ## 
+*(add to III.1.8.1.3 Merging stack states)*
+
+Merging a returnable managed pointer with an ordinary (that is, non-returnable) managed pointer to the same type results in a non-returnable managed pointer to that type.
+
+## Returnable local slots ##
+CLI needs to add a notion of a returnable local slots.
+-   a local byref slot may be marked as returnable. 
+-	`STLOC` into returnable slots is verifiable only when the value is a returnable managed pointer. 
+-	`LDLOC` from a returnable slot produces a returnable managed pointer.
+
+The rationale for introducing returnable locals is to allow for a single pass verification analysis. 
+While, in theory, it may be possible that the �returnable� property could be inferred via some form of fixed point flow analysis, it would make verification substantially more complex and expensive and would not retain the single-pass property.
+
+## Metadata encoding of returnable local slots ## 
+
+Local slots can be marked as returnable by applying `modopt[System.Runtime.CompilerServices.IsReturnableAttribute]` in the local signature.
```

Comment:
Is there a particular reason to use an `Attribute` suffix here? This will serve similar purpose as `System.Runtime.CompilerServices.IsVolatile` or `IsConst`. Those are not attributes either.

Reply:
No, no reason to be an attribute. 

In the parallel case with readonly references, it seems appropriate to use `IsReadOnlyAttribute` since that is used in other places to express readonliness.
I could not think of a similar existing type for "returnable", so I suggest `IsReturnableAttribute` just for similarity, but you are right - it does not need to be an attribute or even a new type. 

Any alternative ideas here would be appreciated, but we need to pick some type.